File: Chapters/Additional_Codes/1D_SWE_CGDG_Wetting_and_Drying/create_rhs.m
%---------------------------------------------------------------------%
%This function computes the LGL grid and elements.
%Written by F.X. Giraldo on 10/2003
%           Department of Applied Maths
%           Naval Postgraduate School
%           Monterey, CA 93943-5216
%---------------------------------------------------------------------%
function rhs = create_rhs(qp,qb,coord,nelem,ngl,nq,wnq,psi,dpsi,h_eps,gravity,delta_nl)

%Initialize
rhs=zeros(2,ngl,nelem);
ps_e=zeros(ngl,1);
pu_e=zeros(ngl,1);
u_e=zeros(ngl,1);
b_e=zeros(ngl,1);
x=zeros(ngl,1);

%Integrate Divergence of Flux (Weak Form)
for ie=1:nelem
   
   %Store Coordinates
   for i=1:ngl
      ps_e(i)=qp(1,i,ie);
      pu_e(i)=qp(2,i,ie);
      b_e(i) =qb(i,ie);
      x(i)   =coord(i,ie);
   end
   
   %Jacobians
   dx=x(ngl)-x(1);
   jac=dx/2;
   ksi_x=2/dx;
   
   %LGL Integration
   for l=1:nq
      wq=wnq(l)*jac;
            
      %Form Derivative
      ps_k=0;
      pu_k=0;
      b_k=0;
      dbdx=0;
      for j=1:ngl
         h_k=psi(j,l);
         dhdx=dpsi(j,l)*ksi_x;
         pu_k=pu_k + pu_e(j)*h_k;
         b_k =b_k  + b_e(j)*h_k;
         ps_k=ps_k + ps_e(j)*h_k;
         dbdx=dbdx + b_e(j)*dhdx;
      end
      
      %-----New
      p_k=ps_k + b_k;
      if (p_k <= h_eps)
          p_k=h_eps;
          %ps_k=p_k - b_k;
          pu_k=0;
      end
      %-----New
      
      flux_p=pu_k;
      flux_pu=( pu_k^2/(p_k) + 0.5*gravity*ps_k^2 )*delta_nl + gravity*b_k*ps_k;
   
      %Form RHS
      for i=1:ngl
         h_i=psi(i,l);
         dhdx_i=dpsi(i,l)*ksi_x;
         rhs(1,i,ie)=rhs(1,i,ie) + wq*dhdx_i*flux_p;
         rhs(2,i,ie)=rhs(2,i,ie) + wq*(dhdx_i*flux_pu + h_i*gravity*ps_k*dbdx);

      end %i
   end %l
end %ie

File: Projects/Project_02_1D_Wave/For_Instructors/julia/RHS_Vector_approach/create_rhs.jl
#---------------------------------------------------------------------#
#This function computes the global Mass and Differentiation Matrices.
#Written by F.X. Giraldo on April 19 2019
#           Department of Applied Mathematics
#           Naval Postgraduate School
#           Monterey; CA 93943-5216
#---------------------------------------------------------------------#
function create_rhs(q,u,M,Dwe,intma,periodicity,DFloat)

    #Get lengths of arrays
    (Np,nelem)=size(intma)
    Npoin=size(q)

    #Store local arrays
    inode=zeros(Int64,Np)
    rhs=zeros(DFloat,Npoin)

    #store flux term
    f=q*u

    #build differentiation matrix contribution
    fe=zeros(DFloat,Np)
    for e=1:nelem
        for i=1:Np
            I=intma[i,e]
            fe[i]=f[I]
        end
        for i=1:Np
            I=intma[i,e]
            for j=1:Np
                rhs[periodicity[I]]+=Dwe[i,j]*fe[j]
            end
        end
    end

    #build flux matrix contribution (numerical flux)
    for e=1:nelem
        L=intma[Np,e]
        ep1=e+1
        if (e == nelem)
            ep1=1
        end
        R=intma[1,ep1]
        flux=0.5*( f[L] + f[R] - u*(q[R] - q[L]) )
        rhs[periodicity[L]]-=flux
        rhs[periodicity[R]]+=flux
    end

    #Build RHS vector
    R=M\rhs

    #Return arrays
    return(R)
end
